File: apps/api/internal/adapter/handler/workout.go
package handler

import (
	"fmt"
	"gogym-api/internal/adapter/dto"
	"gogym-api/internal/util"
	"log/slog"
	"net/http"

	dom "gogym-api/internal/domain/entities"
	wu "gogym-api/internal/application/workout"

	"github.com/labstack/echo/v4"
)

type WorkoutHandler struct {
	wu wu.WorkoutUseCase
}

func NewWorkoutHandler(wu wu.WorkoutUseCase) *WorkoutHandler {
	return &WorkoutHandler{
		wu: wu,
	}
}

func (h *WorkoutHandler) GetWorkoutRecords(c echo.Context) error {
	ctx := c.Request().Context()
	slog.InfoContext(ctx, "GetWorkoutRecords Handler")

	userID, ok := c.Get("user_id").(string)
	if !ok || userID == "" {
		slog.ErrorContext(ctx, "User ID not found in context")
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "Unauthorized"})
	}

	dateStr := c.QueryParam("date")
	date, err := util.ParseJSTDateOrToday(dateStr)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to parse date", "dateStr", dateStr, "error", err)
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "Invalid date format"})
	}

	response, err := h.wu.GetWorkoutRecords(ctx, userID, date)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to get workout records", "userID", userID, "date", date, "error", err)
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, response)
}

func (h *WorkoutHandler) CreateWorkoutRecord(c echo.Context) error {
	ctx := c.Request().Context()
	slog.InfoContext(ctx, "CreateWorkoutRecord Handler")

	userID, ok := c.Get("user_id").(string)
	if !ok || userID == "" {
		slog.ErrorContext(ctx, "User ID not found in context")
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "Unauthorized"})
	}

	var req dto.WorkoutRecordDTO
	err := c.Bind(&req)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to bind request", "error", err)
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "Invalid request format"})
	}

	// dto → domain変換
	domainRecord, err := dto.WorkoutRecordDTOToDomain(&req)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to convert DTO to domain model", "error", err)
		return c.JSON(http.StatusBadRequest, map[string]string{"error": fmt.Sprintf("Invalid request data: %v", err)})
	}

	// Set userID
	domainRecord.UserID = dom.ULID(userID)

	// gym_name から gym_id を解決（gym_name優先、なければgym_idをそのまま使用）
	if req.GymName != nil && *req.GymName != "" {
		gymID, err := h.wu.ResolveGymIDFromName(ctx, userID, *req.GymName)
		if err != nil {
			slog.ErrorContext(ctx, "Failed to resolve gym_id from gym_name", "userID", userID, "gymName", *req.GymName, "error", err)
			return c.JSON(http.StatusInternalServerError, map[string]string{"error": "Failed to resolve gym"})
		}
		domainRecord.GymID = &gymID
	}

	err = h.wu.CreateWorkoutRecord(ctx, *domainRecord)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to create workout record", "userID", userID, "error", err)
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusCreated, map[string]string{"message": "Workout record created successfully"})
}

func (h *WorkoutHandler) UpdateWorkoutRecord(c echo.Context) error {
	ctx := c.Request().Context()
	slog.InfoContext(ctx, "UpdateWorkoutRecord Handler")

	userID, ok := c.Get("user_id").(string)
	if !ok || userID == "" {
		slog.ErrorContext(ctx, "User ID not found in context")
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "Unauthorized"})
	}

	// パスパラメータから record ID を取得（使用しない可能性もあるが、一応取得）
	// recordID := c.Param("id")

	var req dto.WorkoutRecordDTO
	err := c.Bind(&req)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to bind request", "error", err)
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "Invalid request format"})
	}

	// dto → domain変換
	domainRecord, err := dto.WorkoutRecordDTOToDomain(&req)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to convert DTO to domain model", "error", err)
		return c.JSON(http.StatusBadRequest, map[string]string{"error": fmt.Sprintf("Invalid request data: %v", err)})
	}

	// Set userID
	domainRecord.UserID = dom.ULID(userID)

	// gym_name から gym_id を解決（gym_name優先、なければgym_idをそのまま使用）
	if req.GymName != nil && *req.GymName != "" {
		gymID, err := h.wu.ResolveGymIDFromName(ctx, userID, *req.GymName)
		if err != nil {
			slog.ErrorContext(ctx, "Failed to resolve gym_id from gym_name", "userID", userID, "gymName", *req.GymName, "error", err)
			return c.JSON(http.StatusInternalServerError, map[string]string{"error": "Failed to resolve gym"})
		}
		domainRecord.GymID = &gymID
	}

	// Upsert を使用（同日同部位の場合は更新、異なる場合は新規作成）
	err = h.wu.UpsertWorkoutRecord(ctx, *domainRecord)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to update workout record", "userID", userID, "error", err)
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, map[string]string{"message": "Workout record updated successfully"})
}

func (h *WorkoutHandler) GetWorkoutParts(c echo.Context) error {
	ctx := c.Request().Context()
	slog.InfoContext(ctx, "GetWorkoutParts Handler")

	userID, ok := c.Get("user_id").(string)
	if !ok || userID == "" {
		slog.ErrorContext(ctx, "User ID not found in context")
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "Unauthorized"})
	}

	parts, err := h.wu.GetWorkoutParts(ctx, userID)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to get workout parts", "userID", userID, "error", err)
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	slog.InfoContext(ctx, "Successfully retrieved workout parts", "userID", userID, "count", len(parts))
	return c.JSON(http.StatusOK, parts)
}

func (h *WorkoutHandler) SeedWorkoutParts(c echo.Context) error {
	ctx := c.Request().Context()
	slog.InfoContext(ctx, "SeedWorkoutParts Handler")

	userID, ok := c.Get("user_id").(string)
	if !ok || userID == "" {
		slog.ErrorContext(ctx, "User ID not found in context")
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "Unauthorized"})
	}

	err := h.wu.SeedWorkoutParts(ctx, userID)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to seed workout parts", "userID", userID, "error", err)
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, map[string]string{"message": "Workout parts seeded successfully"})
}

func (h *WorkoutHandler) CreateWorkoutExercise(c echo.Context) error {
	ctx := c.Request().Context()
	slog.InfoContext(ctx, "CreateWorkoutExercise Handler")

	userID, ok := c.Get("user_id").(string)
	if !ok || userID == "" {
		slog.ErrorContext(ctx, "User ID not found in context")
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "Unauthorized"})
	}

	var req dto.CreateWorkoutExerciseRequest
	err := c.Bind(&req)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to bind request", "error", err)
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "Invalid request format"})
	}

	err = h.wu.CreateWorkoutExercise(ctx, userID, req.Exercises)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to create workout exercises", "userID", userID, "error", err)
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusCreated, map[string]string{"message": "Workout exercises created successfully"})
}

func (h *WorkoutHandler) DeleteWorkoutExercise(c echo.Context) error {
	ctx := c.Request().Context()
	slog.InfoContext(ctx, "DeleteWorkoutExercise Handler")

	userID, ok := c.Get("user_id").(string)
	if !ok || userID == "" {
		slog.ErrorContext(ctx, "User ID not found in context")
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "Unauthorized"})
	}

	exerciseIDStr := c.Param("id")
	var exerciseID int64
	if _, err := fmt.Sscanf(exerciseIDStr, "%d", &exerciseID); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "Invalid exercise ID format"})
	}

	err := h.wu.DeleteWorkoutExercise(ctx, userID, exerciseID)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to delete workout exercise", "userID", userID, "exerciseID", exerciseID, "error", err)
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, map[string]string{"message": "Workout exercise deleted successfully"})
}

func (h *WorkoutHandler) GetLastWorkoutRecord(c echo.Context) error {
	ctx := c.Request().Context()
	slog.InfoContext(ctx, "GetLastWorkoutRecord Handler")

	userID, ok := c.Get("user_id").(string)
	if !ok || userID == "" {
		slog.ErrorContext(ctx, "User ID not found in context")
		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "Unauthorized"})
	}

	exerciseIDStr := c.Param("id")
	var exerciseID int64
	if _, err := fmt.Sscanf(exerciseIDStr, "%d", &exerciseID); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "Invalid exercise ID format"})
	}

	response, err := h.wu.GetLastWorkoutRecord(ctx, userID, exerciseID)
	if err != nil {
		slog.ErrorContext(ctx, "Failed to get last workout record", "userID", userID, "exerciseID", exerciseID, "error", err)
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, response)
}

File: apps/web/lib/bff/workout.ts
import type {
  WorkoutFormDTO,
  WorkoutPartDTO,
  ExerciseDTO,
} from "@/app/workout/_lib/types";

/**
 * ワークアウト部位をシード（初期データ作成）
 */
export const seedWorkoutParts = async (): Promise<void> => {
  const res = await fetch("/api/workouts/seed", {
    method: "POST",
    cache: "no-store",
  });
  if (!res.ok) throw new Error("failed to seed workout parts");
};

/**
 * ワークアウト部位リストを取得
 */
export const getWorkoutParts = async (): Promise<WorkoutPartDTO[]> => {
  const res = await fetch("/api/workouts/parts", { cache: "no-store" });
  if (!res.ok) return [];
  return res.json();
};

/**
 * ワークアウト記録を取得
 */
export const getWorkoutRecord = async (params?: {
  date?: string;
  partId?: number | null;
}): Promise<WorkoutFormDTO> => {
  const q = new URLSearchParams();
  if (params?.date) q.set("date", params.date);
  if (params?.partId) q.set("part_id", String(params.partId));

  const url = q.toString()
    ? `/api/workouts/records?${q.toString()}`
    : "/api/workouts/records";
  const res = await fetch(url, { cache: "no-store" });
  if (!res.ok) throw new Error("failed to fetch workout record");
  return res.json();
};

/**
 * ワークアウト記録を作成
 */
export const createWorkoutRecord = async (
  body: any,
): Promise<{ ok: boolean; error: string | null }> => {
  const res = await fetch("/api/workouts/records", {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
    cache: "no-store",
  });
  return { ok: res.ok, error: res.ok ? null : "保存に失敗しました" };
};

/**
 * ワークアウト記録を更新
 */
export const updateWorkoutRecord = async (
  id: number,
  body: any,
): Promise<{ ok: boolean; error: string | null }> => {
  const res = await fetch(`/api/workouts/records/${id}`, {
    method: "PUT",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
    cache: "no-store",
  });
  return { ok: res.ok, error: res.ok ? null : "更新に失敗しました" };
};

/**
 * 前回のエクササイズ記録を取得
 */
export const getLastExerciseRecord = async (
  exerciseID: number,
): Promise<ExerciseDTO | null> => {
  if (!exerciseID) {
    return null;
  }

  const res = await fetch(
    `/api/workouts/exercises?id=${exerciseID}&action=last`,
    {
      cache: "no-store",
    },
  );

  if (!res.ok || res.status === 204) {
    return null;
  }

  const data = await res.json();
  return data || null;
};

/**
 * ワークアウト種目を一括作成/更新（Upsert）
 */
export const upsertWorkoutExercises = async (
  exercises: Array<{
    id?: number;
    name: string;
    workout_part_id: number | null;
  }>,
): Promise<{ ok: boolean; error: string | null; data?: any }> => {
  const res = await fetch("/api/workouts/exercises", {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ exercises }),
    cache: "no-store",
  });

  const data = res.ok ? await res.json().catch(() => null) : null;
  return {
    ok: res.ok,
    error: res.ok ? null : "種目の登録に失敗しました",
    data,
  };
};

/**
 * ワークアウト種目を削除
 */
export const deleteWorkoutExercise = async (
  exerciseID: number,
): Promise<{ ok: boolean; error: string | null }> => {
  const res = await fetch(`/api/workouts/exercises?id=${exerciseID}`, {
    method: "DELETE",
    cache: "no-store",
  });

  return { ok: res.ok, error: res.ok ? null : "種目の削除に失敗しました" };
};
